docs/Configuration.md
# Hardware Configuration

Before building and flashing the firmware you must review and possibly adjust the hardware configuration.

Here is a short checklist of things to consider:

* What flight controller board will you be using? [Several boards are supported.](Targets.md)
* What type of receiver will you be using?
    * On-board SPI AFHDS/AFHDS2A is available when using the "HappyModel Crazybee F4 Lite 1S" flight controller as found on the Mobula6 (this is a `MATEKF411RX` target board) and also with the `CRAZYBEEF3FS` target.
    * External SPI Bayang is available using various [transceiver modules.](Transceiver.md).
        * I've had great success with the [Play F4 board and an NRF24L01](PlayF4_NRF24L01.md). That board also works quite well with XN297L modules harvested from toy transmitters.
* Configure your aux channels: throttle kill switch (similar to an arm switch), level mode enable/disable, turtle mode, motor beeps, etc
* Will you be using RPM filtering?
    * If so, you'll want to make sure your ESCs are flashed with either [JESC](https://jflight.net/), [JazzMaverick](https://github.com/JazzMaverick/BLHeli/tree/JazzMaverick-patch-1/BLHeli_S%20SiLabs) or [Bluejay](https://github.com/mathiasvr/bluejay).

## Specifying the flight controller

Several [flight controller targets](Targets.md) are currently supported (`MATEKF411RX`, `CRAZYBEEF3FS`, `NOX`, `OMNIBUSF4`). The `MATEKF411RX` target can support an on-board SPI transceiver (only FlySky AFHDS/AFHDS2A is supported at this time). The `CRAZYBEEF3FS` target also supports on-board SPI FlySky AFHDS/AFHDS2A. 

I'm actively using many of these target boards for my whoops and micros with the exception of the `OMNIBUSF4`, that target was only used for development as it provides SWD solder pads. Support for the `OMNIBUS` (F3 processor) board was started but has yet to be completed.

You will not need to edit any source files to specify your target board. Instead you just need to make note of the target name
(such as `NOX` or `MATEKF411RX`). You'll need to know this when you build the firmware as described in the [Develop](Develop.md) section of this document.

## Specifying the receiver

If you wish to use an **external** SPI RX transceiver [4 different modules](Transceiver.md) are supported. IBUS support is also available; I've successfully used the FlySky FS-RX2A receiver with IBUS on the `NOX` and `OMNIBUSF4` targets. And if you're using that Mobula6 board or the CrazyBeeF3FS board then you can use the on-board SPI AFHDS/AFHDS2A receiver.

You will edit the `_my_config.h` source file to specify your receiver option. Using a text editor (I suggest Visual Studio Code) open the file and look near the top of the file for something that looks like this:

```c++
//------------------------------------------------------------------------------
// RX protocol and configuration
// Enable only one of the following defines
//------------------------------------------------------------------------------
//#define RX_SILVERLITE_BAYANG_PROTOCOL   // Enable SilverLite SPI Transceiver RX implementation
//#define RX_IBUS // Enable IBUS protocol support on a USART RX pin, double-check rx_ibus.cpp and define one of: FLYSKY_i6_MAPPING, TURNIGY_EVOLUTION_MAPPING
//#define RX_FLYSKY     // Enable FlySky (AFHDS) SPI transceiver implementation
#define RX_FLYSKY2A   // Enable FlySky (AFHDS-2A) SPI transceiver implementation
```

The lines that start with `//` are "commented out", meaning they don't do anything.
You basically want to comment out 2 of the 3 choices. The example above enables
the FlySky (AFHDS/AFHDS2A) SPI transceiver option. If instead you want to use an external RX module
with IBUS protocol then you'd edit that text to look like this intead:

```c++
//------------------------------------------------------------------------------
// RX protocol and configuration
// Enable only one of the following defines
//------------------------------------------------------------------------------
//#define RX_SILVERLITE_BAYANG_PROTOCOL   // Enable SilverLite SPI Transceiver RX implementation
#define RX_IBUS // Enable IBUS protocol support on a USART RX pin, double-check rx_ibus.cpp and define one of: FLYSKY_i6_MAPPING, TURNIGY_EVOLUTION_MAPPING
//#define RX_FLYSKY     // Enable FlySky (AFHDS) SPI transceiver implementation
//#define RX_FLYSKY2A   // Enable FlySky (AFHDS-2A) SPI transceiver implementation
```

If you choose the `RX_SILVERLITE_BAYANG_PROTOCOL` option (an external SPI transceiver module
configured to run Bayang protocol), then it would look like the snippet below, *plus* you'll
also want to review and possibly edit the text immediately below that.

```c++
//------------------------------------------------------------------------------
// RX protocol and configuration
// Enable only one of the following defines
//------------------------------------------------------------------------------
#define RX_SILVERLITE_BAYANG_PROTOCOL   // Enable SilverLite SPI Transceiver RX implementation
//#define RX_IBUS // Enable IBUS protocol support on a USART RX pin, double-check rx_ibus.cpp and define one of: FLYSKY_i6_MAPPING, TURNIGY_EVOLUTION_MAPPING
//#define RX_FLYSKY     // Enable FlySky (AFHDS) SPI transceiver implementation
//#define RX_FLYSKY2A   // Enable FlySky (AFHDS-2A) SPI transceiver implementation

//------------------------------------------------------------------------------
// When using RX_SILVERLITE_BAYANG_PROTOCOL you must specify which transceiver
// module you're using and whether or not you're using 3-wire SPI or 4-wire SPI.
//
// Note:  The software SPI pins used for interfacing with the module are defined 
// in file: trx_spi_config.h
//------------------------------------------------------------------------------
#ifdef RX_SILVERLITE_BAYANG_PROTOCOL

// Define only one of the TRX_??? values below
#define TRX_NRF
//#define TRX_XN297
//#define TRX_XN297L
//#define TRX_LT8900
```

See the line that says `// Define only one of the TRX_??? values below`?
You'll do something similar here, you'll comment out 3 of the 4 choices.
The snippet above has enabled only the `TRX_NRF` option. This means you'll be
using an NRF24L01 transceiver module for your receiver (and also to transmit back
telemetry).

### Configuring SPI

When using an external SPI transceiver module you will want to review the pin assignments that
connect the STM32 processor to your external module. 

Examine the `trx_spi_config.h` header file and it should be commented well enough to figure out how
things are hooked up and how you can change things if needed.

The [Play F4 and NRF24L01 section](PlayF4_NRF24L01.md) provides a pretty good walkthru of how to
connect an NRF24L01 module to a "Play F4" flight controller board.


## Configuring Rates and Expo
The next section of the `_my_config.h` file is where you'll define your rates and expo.

```c++
//------------------------------------------------------------------------------
// Rates
//------------------------------------------------------------------------------

// rate in deg/sec for acro mode
#define MAX_RATE            800
#define MAX_RATEYAW         675

#define LEVEL_MAX_ANGLE     80
#define LEVEL_MAX_RATE      900

#define LOW_RATES_MULTI     0.65

//------------------------------------------------------------------------------
// Expo
//  Allowed values are 0.00 to 1.00. A value of 0 means no expo applied
//  The higher the value, the less sensitive near center
//------------------------------------------------------------------------------

#define ACRO_EXPO_ROLL      0.85
#define ACRO_EXPO_PITCH     0.85
#define ACRO_EXPO_YAW       0.26

#define ANGLE_EXPO_ROLL     0.55
#define ANGLE_EXPO_PITCH    0.55
#define ANGLE_EXPO_YAW      0.30

```

## Configuring PID terms

The next section in `_my_config.h` is where you'll define your PID terms for acro and angle (level) modes.

```c++
//------------------------------------------------------------------------------
// PID term overrides
//------------------------------------------------------------------------------
                        //  Roll    Pitch   Yaw
#define     ACRO_P      {   .040,   .040,   .01     };
#define     ACRO_I      {   .250,   .250,   .50     };
#define     ACRO_D      {   .035,   .035,   .0      };

// Angle mode P and D terms
#define     ANGLE_P1    10.
#define     ANGLE_D1    3.0
```

The values shown here are decent starting points for 65mm or 75mm whoops as well as 2.5" and 3" micros using 0802, 1102 or 1103 motors.

You can always fine tune them in the field using stick gestures and the on screen display.

*TODO*: Document the PID tuning feature


## More and more
The rest of the `_my_config.h` file is where you configure even more options that 
are available to you. I won't go into great detail on them but here's a list of what to expect:

* `RPM_FILTER` - If defined, this will enable RPM filtering. By default this is defined for all targets. Be sure you've already flashed your ESCs with firmware that supports RPM filtering.
* `LOOPTIME` - By default the looptime is configured for 250us (4k loop). You really can't go higher unless you disable RPM filtering and/or overclock. This value works very well with the targets supported by SilverLite.
* Various filters. Read the comments in the source file for more info
* Switches/Channels - What aux channels (switches on your transmitter) are supported and what they should do. This section is used to map 4 features (`THROTTLE_KILL_SWITCH`, `LEVELMODE`, `MOTOR_BEEPS_CHANNEL` and `RATES`) to some Bayang channels. It requires you know what Bayang channels your TX is setup to use when it transmits with the Bayang protocol.
    * Note: If you're using IBUS or AFHDS/AFHDS2 protocol you will instead need to edit some code in `rx_flysky.cpp` or `rx_ibus.cpp`. I apologize in advance for not documenting this further. 

### Even more config

If you have need to tweak the DSHOT configuration, or motor idle offset or motor pole count then you'll want to review and edit the `_myHardware.h` file.



Core/Src/config.h
#include "defines.h"

// PID values in pid.c

// rate in deg/sec for acro mode
#define MAX_RATE 1800
#define MAX_RATEYAW 1800

// max angle for level mode
#define LEVEL_MAX_ANGLE 80

// max rate used by level pid (limit)
#define LEVEL_MAX_RATE 1800

#define LOW_RATES_MULTI 0.5

// Change this factor to get a correct battery voltage.
#define ADC_SCALEFACTOR 11.111 // 11.0 for an ideal 10k/1k voltage divider

// Make sure to understand CELL_COUNT_UNSCALED in battery.c before enabling this.
#define BATTERY_CELL_COUNT_DETECTION

// Do not start software if battery is too low. Flashes 2 times repeatedly at startup.
// #define STOP_LOWBATTERY // If below 3.3 Volt

// If enabled, start LED blinking at low battery voltage
#define WARN_ON_LOW_BATTERY 3.6 // Volt

// Voltage hysteresis for WARN_ON_LOW_BATTERY
#define VOLTAGE_HYSTERESIS 0.10 // Volt

// compensation for battery voltage vs throttle drop
#define VDROP_FACTOR 0.7
// calculate above factor automatically
#define AUTO_VDROP_FACTOR

// lower throttle when battery below threshold
#define LVC_LOWER_THROTTLE
#define LVC_LOWER_THROTTLE_VOLTAGE 3.30
#define LVC_LOWER_THROTTLE_VOLTAGE_RAW 2.70
#define LVC_LOWER_THROTTLE_KP 3.0

// MPU-60x0 on-chip Gyro LPF filter frequency
// gyro filter 0: 256 Hz, delay 0.98 ms (use this to get 8k gyro update frequency)
// gyro filter 1: 188 Hz, delay 1.9 ms
// gyro filter 2: 98 Hz, delay 2.8 ms
// gyro filter 3: 42 Hz, delay 4.8 ms
#define GYRO_LOW_PASS_FILTER 0

// Gyro Notch Filters

#define RPM_FILTER // requires DSHOT_DMA_BIDIR in hardware.h -- also ensure MOTOR_POLE_COUNT in drv_dshot_bidir.c is correct
#define RPM_FILTER_HZ_MIN 100
#define RPM_FILTER_2ND_HARMONIC false // note, that there are 12 notch filters (4 motors * 3 axes) per harmonic
#define RPM_FILTER_3RD_HARMONIC true
#define RPM_FILTER_Q 6 // -3dB bandwidth = f0 / Q -- but a higher Q also results in a longer settling time

// #define BIQUAD_NOTCH_A_HZ 260 // Dalprop Cyclone T5249C
// #define BIQUAD_NOTCH_A_Q 6

// #define BIQUAD_NOTCH_B_HZ 300 // T-Motor T5147
// #define BIQUAD_NOTCH_B_Q 6

// #define BIQUAD_NOTCH_C_HZ 380
// #define BIQUAD_NOTCH_C_Q 6

// Dynamic Gyro first and second order LPFs

#define GYRO_LPF_1ST_HZ_BASE 120 // Filter frequency at zero throttle.
#define GYRO_LPF_1ST_HZ_MAX 120 // A higher filter frequency than loopfrequency/2.4 causes ripples.
#define GYRO_LPF_1ST_HZ_THROTTLE 0.25 // MAX reached at 1/4 throttle.

// #define GYRO_LPF_2ND_HZ_BASE 240 //* ( aux[ FN_INVERTED ] ? 0.75f : 1.0f )
// #define GYRO_LPF_2ND_HZ_MAX 240
// #define GYRO_LPF_2ND_HZ_THROTTLE 0.25

// Static Gyro first order LPF
//#define GYRO_LPF_1ST_HZ 240

// Additional static Gyro first order LPF on yaw only
//#define GYRO_YAW_LPF_1ST_HZ 240

// Dynamic D-Term second order LPF (cannot be turned off)
#define DTERM_LPF_2ND_HZ_BASE 60 //* ( aux[ FN_INVERTED ] ? 0.75f : 1.0f )
#define DTERM_LPF_2ND_HZ_MAX 60
#define DTERM_LPF_2ND_HZ_THROTTLE 0.5

// Additional static D-Term first order LPF
//#define DTERM_LPF_1ST_HZ 120

// Whether to use Bessel type filter for D-Term instead of PT2.
//#define DTERM_BESSEL_FILTER

// If enabled, the D-Term filter uses the filtered gyro signal from above. (Notch filters are always applied.)
//#define CASCADE_GYRO_AND_DTERM_FILTER

// Switch function selection

#define RATES DEVO_CHAN_9 // LOW_RATES_MULTI gets applied when RATES is 0.

#define LEVELMODE DEVO_CHAN_10 // Set this to CH_ON for level mode always on. Comment it out for just acro.

#define LEDS_ON DEVO_CHAN_7

// To stop the motors on ground a switch on the remote control is necessary.
#define THROTTLE_KILL_SWITCH DEVO_CHAN_5

// enable inverted (3D) flight code
#define INVERTED_ENABLE // goes together with BIDIRECTIONAL in drv_dshot.c and drv_dshot_dma.c
#define FN_INVERTED DEVO_CHAN_6
// #define LEVEL_MODE_INVERTED_ENABLE // be careful when enabling this

// Two switchable channels via gestures: CH_AUX1 and CH_AUX2
// Channel CH_AUX1 changed via gestures LLU -> 1 and LLD -> 0
// Channel CH_AUX2 changed via gestures RRU -> 1 and RRD -> 0
// #define AUX1_START_ON // CH_AUX1 channel starts on if this is defined, otherwise off.
// #define AUX2_START_ON // CH_AUX2 channel starts on if this is defined, otherwise off.

// lost quad beeps using motors (60 sec timeout or via channel)
#define MOTOR_BEEPS
#define MOTOR_BEEPS_CHANNEL DEVO_CHAN_12

// Send maximum looptime in the telemetry data.
#define DISPLAY_MAX_USED_LOOP_TIME_INSTEAD_OF_RX_PACKETS
// Send PID values in the telemetry data.
#define DISPLAY_PID_VALUES

// Radio module and protocol selection (only Bayang protocol implemented)
#define RX_NRF24_BAYANG_TELEMETRY // For nRF24L01+ radio module
// #define RX_BAYANG_PROTOCOL_TELEMETRY // For XN297 radio module harvested from toy TX

#define RADIO_XN297 // also enable SOFTSPI_4WIRE in hardware.h
// #define RADIO_XN297L // also enable SOFTSPI_3WIRE in hardware.h

#define TX_POWER 3 // 0 .. 3 (try 1 when using an nRF24L01+PA+LNA module)

// led brightness 0 .. 15 (used for solid lights only)
#define LED_BRIGHTNESS 15

// PID tuning by gestures and/or stick position
//#define PID_GESTURE_TUNING
#define PID_STICK_TUNING
#define COMBINE_PITCH_ROLL_PID_TUNING

// a filter which makes throttle feel faster (aka Throttle Boost) (not active in LOW_RATES or lowbatt)
#define THROTTLE_TRANSIENT_COMPENSATION_FACTOR 3.0

// Compensate throttle for sagging battery voltage
//#define THROTTLE_VOLTAGE_COMPENSATION

// For smoother motor reversing in 3D flight
#define THROTTLE_REVERSING_KICK 0.15 // 4S
#define THROTTLE_REVERSING_DEADTIME 20000 // 20 ms (increase this in case of over-propped motors)
//#define THROTTLE_STARTUP_KICK 0.10

// Continue stick movement with the current stick velocity in case of lost packets
#define RX_PREDICTOR
// Add linear interpolation between the otherwise 5 ms staircase steps of the RX signal
#define RX_SMOOTHING

// Betaflight like mix scaling (aka Airmode)
#define MIX_SCALING
// Mix increasing yields a more crisp response but also a more jumpy quad at low RPM
#define ALLOW_MIX_INCREASING
// A higher value means a shorter active increasing period (shorter bouncy period)
#define TRANSIENT_MIX_INCREASING_HZ 2.0
// Can be used to limit maximum motor RPM, i.e. tone down a too fast quad.
#define MIX_RANGE_LIMIT 1.0f // aux[ DEVO_CHAN_11 ] ? 0.75f : 1.0f

// Adjust IDLE_OFFSET so that the motors still spin reliably at zero throttle.
#define IDLE_OFFSET 0.02 // 4S
// Use a square root motor curve to counteract thrust ~ RPM^2
//#define THRUST_LINEARIZATION 0.33f // 0.0 .. no compensation, 1.0 .. full square root curve
#define ALTERNATIVE_THRUST_LINEARIZATION 0.6 // different shape from the above curve; do not enable both together

// A deadband can be used to eliminate stick center jitter and non-returning to exactly 0.
#define STICKS_DEADBAND 0.02f

// throttle direct to motors for thrust measure
// #define MOTORS_TO_THROTTLE

// throttle direct to motors for thrust measure as a flight mode
#define MOTORS_TO_THROTTLE_MODE CH_AUX1

// Compensate PID values for sagging battery voltage
#define PID_VOLTAGE_COMPENSATION

// Invert yaw pid. Necessary when spinning props outwards.
#define INVERT_YAW_PID

// Rotate I-term vector for a stable yaw axis (aka iTerm Rotation)
#define PID_ROTATE_ERRORS

// Remove roll and pitch bounce back after flips (aka iTerm Relax)
#define TRANSIENT_WINDUP_PROTECTION
// Remove bounce back when quickly stopping a roll/pitch/yaw movement (but it is mostly there for yaw)
#define DYNAMIC_ITERM_RESET

// Feed fast roll/pitch-stick changes directly to the motors to give a snappier response
// 0.0f (or commented out) equates D-term on measurement, 1.0f equates D-term on error.
//#define FEED_FORWARD_STRENGTH 1.0f
//#define SMART_FF
// Feedforward for yaw. It's an absolute value, not related to the 0 .. 1 from above.
//#define FEED_FORWARD_YAW 0.2f

// Loop time in us
#define LOOPTIME 250

// Failsafe time in us. Sets stick inputs to zero after FAILSAFETIME no RX signal. Keeps quad stabilized.
#define FAILSAFETIME 150000 // 0.15 seconds
// Motors failsafe time in us. Shuts motors down after additional MOTORS_FAILSAFETIME.
#define MOTORS_FAILSAFETIME 3000000 // 3 seconds

// Gyro orientation:
// The expected orientation is with the dot on the chip in the front-left corner.
// Specify by how much you have to rotate the board so that the dot is front left.
// The rotations are performed in order and cumulated.
// Note, the motors don't get rotated, so they have to be referenced to the new gyro position.
//#define SENSOR_ROTATE_45_CCW
//#define SENSOR_ROTATE_45_CW
//#define SENSOR_ROTATE_90_CW
#define SENSOR_ROTATE_90_CCW
//#define SENSOR_ROTATE_180
//#define SENSOR_INVERT // Necessary if the gyro is mounted upside down. For an inverted gyro,
// the expected orientation is with the dot on the chip in the front-right corner.

// Motor order
#define MOTOR_BL 2
#define MOTOR_FL 1
#define MOTOR_BR 4
#define MOTOR_FR 3

// For BIDIRECTIONAL motor direction, the motors can be reversed below.
#define REVERSE_MOTOR_BL false
#define REVERSE_MOTOR_FL false
#define REVERSE_MOTOR_BR false
#define REVERSE_MOTOR_FR false
// Disable the check for known gyro that causes the 4 times LED flash.
#define GYRO_CHECK

// Disable the check for development without RX module (3 times LED flash).
#define RADIO_CHECK

// Logs various information to an externally connected OpenLager logger
//#define BLACKBOX_LOGGING


#ifdef USE_SILVERLITE
// Place your customizations into the following included header file rather
// than directly modifying this file. It will make merging/updating easier.
// It also makes it cleaner/easier to support multiple targets.
#include "_my_config.h"
#endif
